module/zfs/spa.c

```diff
@@ -6469,6 +6478,10 @@ spa_sync(spa_t *spa, uint64_t txg)
 	spa->spa_syncing_txg = txg;
 	spa->spa_sync_pass = 0;
 
+	mutex_enter(&spa->spa_alloc_lock);
+	VERIFY0(avl_numnodes(&spa->spa_alloc_tree));
+	mutex_exit(&spa->spa_alloc_lock);
```

Comment:
This feels pretty heavy handed for a release build.  I'd be tempted to wrap it in ZFS_DEBUG but there's no point diverging here unless it's actually a problem.